hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/ContentSummaryProcessor.java

```diff
@@ -0,0 +1,119 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.azurebfs.AzureBlobFileSystemStore;
+import org.apache.hadoop.fs.azurebfs.utils.ABFSContentSummary;
+
+public class ContentSummaryProcessor {
+  private final AtomicLong fileCount = new AtomicLong(0L);
+  private final AtomicLong directoryCount = new AtomicLong(0L);
+  private final AtomicLong totalBytes = new AtomicLong(0L);
+  private final AtomicInteger numTasks = new AtomicInteger(0);
+  private final AzureBlobFileSystemStore abfsStore;
+  private static final int NUM_THREADS = 16;
+  private final ExecutorService executorService = new ThreadPoolExecutor(1,
+      NUM_THREADS, 5, TimeUnit.SECONDS, new SynchronousQueue<>());
+  private final CompletionService<Void> completionService = new ExecutorCompletionService<>(
+      executorService);
+  private final LinkedBlockingQueue<FileStatus> queue = new LinkedBlockingQueue<>();
+  private static final Logger LOG = LoggerFactory.getLogger(ContentSummaryProcessor.class);
+  private static final int POLL_TIMEOUT = 100;
+
+  public ContentSummaryProcessor(AzureBlobFileSystemStore abfsStore) {
+    this.abfsStore = abfsStore;
+  }
+
+  public ABFSContentSummary getContentSummary(Path path)
+          throws IOException, ExecutionException, InterruptedException {
+    try {
+      processDirectoryTree(path);
+      while (!queue.isEmpty() || numTasks.get() > 0) {
+        LOG.debug("FileStatus queue size = {}, number of submitted unfinished tasks = {}, active thread count = {}",
+                queue.size(), numTasks, ((ThreadPoolExecutor) executorService).getActiveCount());
+        try {
+          completionService.take().get();
+        } finally {
+          numTasks.decrementAndGet();
+        }
+      }
+    } finally {
+      executorService.shutdownNow();
+    }
+
+    return new ABFSContentSummary(totalBytes.get(), directoryCount.get(),
+        fileCount.get(), totalBytes.get());
+  }
+
+  private void processDirectoryTree(Path path)
+      throws IOException, InterruptedException {
+    FileStatus[] fileStatuses = abfsStore.listStatus(path);
+
+    for (FileStatus fileStatus : fileStatuses) {
+      if (fileStatus.isDirectory()) {
+        queue.put(fileStatus);
+        processDirectory();
+        conditionalSubmitTaskToExecutor();
+      } else {
+        processFile(fileStatus);
+      }
+    }
+  }
+
+  private void processDirectory() {
+    directoryCount.incrementAndGet();
+  }
+
+  private void processFile(FileStatus fileStatus) {
+    fileCount.incrementAndGet();
+    totalBytes.addAndGet(fileStatus.getLen());
+  }
+
+  private synchronized void conditionalSubmitTaskToExecutor() {
+    if (!queue.isEmpty() && numTasks.get() < NUM_THREADS) {
+      numTasks.incrementAndGet();
```

Comment:
See if there is a method for simple get, since you are not using the return value

Reply:
Can't find any op without returning, will have to stick to this